books/urls.py

```diff
@@ -11,6 +11,21 @@
         login_required(views.BookUpdate.as_view()),
         name="book_update",
     ),
+    path(
+        "<int:pk>/request_item/<int:userid>",
```

Comment:
You have put the "holder" of the book as part of the url here, similar to the book id.

This possibly causes some design-level concerns to be considered:

- 1 Security. The front-end tells us who the holder is. A person alters the form and can now request the book from anyone. So Mallory wants to request a book from Alice, but instead alters the form so that the userid in the url refers to his friend, Chuck. Chuck can then confirm on behalf of Alice.
- 2 Consistency. The front-end might be outdated. Bob may try to request the book from Alice, but Eve has already requested the book and Alice has already confirmed. So now the holder is actually Eve, but the outdated front-end will still request it from Alice.
- 3 "Restfull-ness". Book id refers to a resource on the server that we can request and perform actions on. User id is not really the resource being accessed, rather it's a parameter to some of the actions. When doing POST requests we generally prefer to put such parameters as part of the body, aka "post data". On a practical level, you may later want to provide additional params to these requests, which would make it inconvenient if all of them went in the url.

The main pain-points are really 1 and 2. These could be solved if you took the responsibility back to the backend, so that your view goes and figures out who the current holder is, without any input  from the front-end.